src/providers/postgres/qgspostgresfeatureiterator.cpp

```diff
@@ -79,28 +86,42 @@ QgsPostgresFeatureIterator::QgsPostgresFeatureIterator( QgsPostgresFeatureSource
   else if ( request.filterType() == QgsFeatureRequest::FilterExpression
             && QSettings().value( "/qgis/compileExpressions", true ).toBool() )
   {
+    //IMPORTANT - this MUST be the last clause added!
     QgsPostgresExpressionCompiler compiler = QgsPostgresExpressionCompiler( source );
 
     if ( compiler.compile( request.filterExpression() ) == QgsSqlExpressionCompiler::Complete )
     {
+      useFallbackWhereClause = true;
+      fallbackWhereClause = whereClause;
       whereClause = QgsPostgresUtils::andWhereClauses( whereClause, compiler.result() );
       mExpressionCompiled = true;
     }
   }
 
-  if ( !mSource->mSqlWhereClause.isEmpty() )
+  if ( !declareCursor( whereClause, false ) )
   {
-    if ( !whereClause.isEmpty() )
-      whereClause += " AND ";
-
-    whereClause += '(' + mSource->mSqlWhereClause + ')';
-  }
-
-  if ( !declareCursor( whereClause ) )
-  {
-    mClosed = true;
-    iteratorClosed();
-    return;
+    if ( useFallbackWhereClause )
+    {
+      //try with the fallback where clause, eg for cases when using compiled expression failed to prepare
+      if ( !declareCursor( fallbackWhereClause ) )
+      {
+        //nothing else to try
+        mClosed = true;
+        iteratorClosed();
+        return;
```

Comment:
I try to avoid `return` in branches because it is easily missed when later adding code like unguarded pointers or other cleanup code at the end of the function.